Gems/ScriptCanvas/Code/Include/ScriptCanvas/AutoGen/ScriptCanvasAutoGenRegistry.h

```diff
@@ -6,89 +6,97 @@
 #include <AzCore/std/string/string.h>
 
 namespace AZ
 {
-    class ReflectContext;
     class ComponentDescriptor;
+    class ReflectContext;
 }
 
-//! Macros to self-register AutoGen node into ScriptCanvas
-#define REGISTER_SCRIPTCANVAS_AUTOGEN_DATA(LIBRARY)\
-    static ScriptCanvas::LIBRARY##DataRegistry s_##LIBRARY##AutoGenDataRegistry;
-#define REGISTER_SCRIPTCANVAS_AUTOGEN_FUNCTION(LIBRARY)\
-    static ScriptCanvas::LIBRARY##FunctionRegistry s_##LIBRARY##AutoGenFunctionRegistry;
-#define REGISTER_SCRIPTCANVAS_AUTOGEN_NODEABLE(LIBRARY)\
-    static ScriptCanvas::LIBRARY##NodeableRegistry s_##LIBRARY##AutoGenNodeableRegistry;
-#define REGISTER_SCRIPTCANVAS_AUTOGEN_GRAMMAR(LIBRARY)\
-    static ScriptCanvas::LIBRARY##GrammarRegistry s_##LIBRARY##AutoGenGrammarRegistry;
-
-//! Not used anymore, keep it for backward compatibility
-#define INIT_SCRIPTCANVAS_AUTOGEN(LIBRARY)
-#define REFLECT_SCRIPTCANVAS_AUTOGEN(LIBRARY, CONTEXT)
-#define GET_SCRIPTCANVAS_AUTOGEN_COMPONENT_DESCRIPTORS(LIBRARY)
-
 namespace ScriptCanvas
 {
-    struct NodeRegistry;
+    class Node;
+}
 
-    class ScriptCanvasRegistry
-    {
-    public:
-        virtual ~ScriptCanvasRegistry() = default;
-        virtual void Init([[maybe_unused]] NodeRegistry* nodeRegistry) {}
-        virtual void Reflect([[maybe_unused]] AZ::ReflectContext* context) {}
-        virtual AZStd::vector<AZ::ComponentDescriptor*> GetComponentDescriptors() { return {}; }
-        void ReleaseDescriptors();
-    protected:
-        AZStd::vector<AZ::ComponentDescriptor*> m_cachedDescriptors;
-    };
-
-    //! AutoGenRegistryManager
-    //! The registry manager contains all autogen functions, nodeables and grammars metadata which will be registered
-    //! for ScriptCanvas
-    class AutoGenRegistryManager
+
+
+#define CONCATENATION_IMPL(a, b) a##b
+#define CONCATENATION(a, b) CONCATENATION_IMPL(a, b)
+
+#define SCRIPTCANVAS_REGISTER_EXTERN(cls) \
+    template <typename T> \
+    extern void RegistrationCall(); \
+    template <> \
+    void RegistrationCall<cls>(); \
+    static const int CONCATENATION(temp, __COUNTER__ ) = \
+        RegistrationHelper::Register<cls>(&RegistrationCall<cls>)
+
+#define SCRIPTCANVAS_REGISTER(cls) \
+    template <> \
+    void RegistrationCall<cls>()
+
+namespace RegistrationHelper
+{
+    template <typename T>
+    inline int Register(void(*f)())
     {
-    public:
-        AutoGenRegistryManager() = default;
-        ~AutoGenRegistryManager();
+        static const int s = [&f]() {
+            f();
+            return 0;
+        }();
+        return s;
+    }
+}
+
+//! Holds the complete list of Script Canvas nodes (grammar and nodeables), all
+//! gems and modules will register their nodes into this model
+class ScriptCanvasModel
+{
+public:
 
-        static AutoGenRegistryManager* GetInstance();
+    static ScriptCanvasModel& Instance();
 
-        //! Init all AutoGen registries
-        static void Init();
+    ScriptCanvasModel() = default;
```

Comment:
Instead of defaulting this constructor, it should have an implementation in the cpp file, where it reads verbose option from either a CVAR or a Settings Registry key.
```suggestion
    ScriptCanvasModel()
    {
        if (auto settingsRegistry = AZ::SettingsRegistry::Get(); settingsRegistry != nullptr)
        {
            constexpr AZStd::string_view VerboseAutoGenRegistration = "/O3DE/Settings/ScriptCanvas/AutoGen/Verbose";
            settingsRegistry->Get(m_verbose, VerboseAutoGenRegistration);
        }
    }
```